components/nexusoperations/statemachine.go

```diff
@@ -460,18 +460,19 @@ var TransitionTimedOut = hsm.NewTransition(
 	},
 )
 
-// Cancel marks the Operation machine as canceled by spawning a child Cancelation machine and transitioning the child to
-// the SCHEDULED state.
+// Cancel marks the Operation machine as canceled by spawning a child Cancelation machine. If the
+// Operation ID is set, then transition the Cancelation machine to the SCHEDULED state. Otherwise,
+// the Cancelation machine will wait the Operation machine transition to the STARTED state.
 func (o Operation) Cancel(node *hsm.Node, t time.Time) (hsm.TransitionOutput, error) {
 	child, err := node.AddChild(CancelationMachineKey, Cancelation{
 		NexusOperationCancellationInfo: &persistencespb.NexusOperationCancellationInfo{},
 	})
 	if err != nil {
-		// This function should be called as part of command/event handling and it should not called more than once.
+		// This function should be called as part of command/event handling and it should not be called
+		// more than once.
 		return hsm.TransitionOutput{}, err
 	}
 	// Operation wasn't started yet, we don't know how to cancel it ATM.
-	// TODO(bergundy): Support cancel-before-started.
 	if o.OperationId == "" {
```

Comment:
This is a bit more accurate.

```suggestion
	if o.State() != enumsspb.NEXUS_OPERATION_STATE_STARTED {
```

Reply:
I cannot change to this. If I check the state instead of the operation ID, there will be a race condition with the Nexus operation `TransitionStarted`. Something like, the Operation state changed to STARTED, but it didn't set the OperationId yet; then this condition passes, the cancelation machine changes to SCHEDULED state, and the cancelation task is processed, but the OperationId field is still missing.

Also, if I only check for the STARTED state, if the Operation machine state is already in a terminal state, then the cancelation machine will stay in UNSPECIFIED state.

Reply:
Hmm... there should never be a race, the two fields are updated atomically.
We should never accept an empty operation ID in the start response but I want to enforce that in the Nexus SDK.

The reason I want to check for started is that operation ID can be not empty after the operation completes and even though we prevent cancelation of completed operations when we handle the workflow command, I'd rather keep the state machine protected here.